learner_pathway_progress/utilities.py

```diff
@@ -0,0 +1,188 @@
+def get_all_learner_pathways(user_id):
+    """Return the all pathways."""
+    user = User.objects.filter(id=user_id).first()
+    client = create_catalog_api_client(user)  # pylint: disable=assignment-from-none
+    learner_pathways_url = urljoin(f"{get_catalog_api_base_url()}/", "learner-pathway/")
+    learner_pathways = None
+    try:
+        response = client.get(learner_pathways_url)
+        response.raise_for_status()
+        learner_pathways = response.json()
+        learner_pathways_count = len(learner_pathways['results'])
+        log.info(
+            f"{PATHWAY_LOGS_IDENTIFIER}: Total Learner Pathways are : {learner_pathways_count}"
+        )
+    except HTTPError as err:
+        log.exception(
+            f"{PATHWAY_LOGS_IDENTIFIER}: Encountered HTTPError {err} while getting all pathways from discovery. "
+        )
+    return learner_pathways
+
+
+def back_fill_learner_pathway_membership_and_progress(pathways):
+    """Update the pathway progress and created membership for user if not already exist."""
+    enrollments_count = EnterpriseCourseEnrollment.objects.all().count()
+    log.info(
+        f"{PATHWAY_LOGS_IDENTIFIER}: Total Enterprise Course Enrollment Count: {enrollments_count}"
+    )
+    for offset in range(0, enrollments_count, CHUNK_SIZE):
+        ent_course_enrollments = EnterpriseCourseEnrollment.objects.all()[
+                                 offset:offset + CHUNK_SIZE].select_related('enterprise_customer_user').only(
+            'course_id',
+            'enterprise_customer_user__user_id',
+        )
+        log.info(
+            f"{PATHWAY_LOGS_IDENTIFIER}: Enterprise Course Enrollment with offset:"
+            f" {offset} and offset chunk size {offset + CHUNK_SIZE}"
+        )
+        for pathway in pathways:
+            pathway_courses = pathway['courses']
+            for enrollment in ent_course_enrollments:
+                user = User.objects.filter(id=enrollment.enterprise_customer_user.user_id).first()
+                pathway_progress_exist = LearnerPathwayProgress.objects.filter(
+                    user=user,
+                    learner_pathway_uuid=pathway['pathway-uuid'],
+                ).exists()
+                if not pathway_progress_exist and enrollment.course_id in pathway_courses:
+
+                    LearnerPathwayMembership.objects.get_or_create(
+                        user=user,
+                        learner_pathway_uuid=pathway['pathway-uuid']
+                    )
+                    pathway_progress = LearnerPathwayProgress.objects.create(
+                        user=user,
+                        learner_pathway_uuid=pathway['pathway-uuid'],
+                    )
+                    pathway_progress.learner_pathway_progress = json.dumps(pathway['snapshot'])
+                    pathway_progress.save()
+                    pathway_progress.update_pathway_progress()
+
+
+def get_pathway_course_and_course_run_keys(step_courses, step_programs, courses):
```

Comment:
it is not returning the list of courses but appending them to the parameter received, is this ok?

Reply:
Yes, because that way we will be returning the value unnecessarily.